cluster-state-service/handler/store/datastore.go

```diff
@@ -110,13 +112,25 @@ func (datastore etcdDataStore) Get(key string) (map[string]string, error) {
 }
 
 // StreamWithPrefix starts a go routine that streams key-value pairs whose keys start with keyPrefix into the channel returned
-func (datastore etcdDataStore) StreamWithPrefix(ctx context.Context, keyPrefix string) (chan map[string]string, error) {
+func (datastore etcdDataStore) StreamWithPrefix(ctx context.Context, keyPrefix string, entityVersion string) (chan map[string]types.Entity, error) {
 	if len(keyPrefix) == 0 {
 		return nil, errors.New("Key prefix cannot be empty while streaming data from datastore by prefix")
 	}
 
-	kvChan := make(chan map[string]string) // go routine datastore.stream() handles closing of this channel
-	go datastore.stream(ctx, keyPrefix, kvChan)
+	// If entity version is specified, verify that it is not out of range.
+	if entityVersion != "" {
+		var revision int64
+		var err error
+		if revision, err = strconv.ParseInt(entityVersion, 10, 64); err != nil {
+			return nil, err
+		}
+		if _, err = datastore.etcdInterface.Get(ctx, keyPrefix, clientv3.WithRev(revision)); err != nil {
```

Comment:
Yup, I needed a way to make sure the revision being specified has not been compacted and does not exceed the latest revision in Etcd. The Etcd watch stream does not throw a compacted or future revision error message unfortunately. It is possible, although fairly unlikely, that a compaction could happen between this check and the start of the stream. However, the bigger issue is that if you specify a future revision, the stream blocks until that revision is created in the future. I've added a TODO to look into a better way of handling this.